pkg/clusteragent/admission/mutate/autoinstrumentation/image_resolver.go

```diff
@@ -0,0 +1,268 @@
+// Assumes rcClient is non-nil.
+func newRemoteConfigImageResolver(rcClient RemoteConfigClient) ImageResolver {
+	resolver := &remoteConfigImageResolver{
+		rcClient:      rcClient,
+		imageMappings: make(map[string]map[string]ResolvedImage),
+	}
+
+	rcClient.Subscribe(state.ProductGradualRollout, resolver.processUpdate)
+	log.Debugf("Subscribed to %s", state.ProductGradualRollout)
+
+	if err := resolver.waitForInitialConfig(5 * time.Second); err != nil {
+		log.Warnf("Failed to load initial image resolution config: %v. Image resolution will be disabled.", err)
+		return newNoOpImageResolver()
+	}
+
+	return resolver
+}
+
+func (r *remoteConfigImageResolver) waitForInitialConfig(timeout time.Duration) error {
+	if currentConfigs := r.rcClient.GetConfigs(state.ProductGradualRollout); len(currentConfigs) > 0 {
+		log.Debugf("Initial configs available immediately: %d configurations", len(currentConfigs))
+		r.updateCache(currentConfigs)
+		return nil
+	}
+
+	// If not immediately available, poll with exponential backoff
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+
+	backoff := 100 * time.Millisecond
+	maxBackoff := 2 * time.Second
+
+	for {
+		select {
+		case <-ctx.Done():
+			return fmt.Errorf("timeout waiting for initial remote config after %v", timeout)
+		case <-time.After(backoff):
+			currentConfigs := r.rcClient.GetConfigs(state.ProductGradualRollout)
+			if len(currentConfigs) > 0 {
+				log.Infof("Loaded initial image resolution config after backoff: %d configurations", len(currentConfigs))
+				r.updateCache(currentConfigs)
+				return nil
+			}
+
+			log.Debugf("Still waiting for initial remote config, retrying in %v", backoff)
+			backoff = min(backoff*2, maxBackoff)
+		}
+	}
+}
+
+// Resolve resolves a registry, repository, and tag to a digest-based reference.
+// Input: "gcr.io/datadoghq", "dd-lib-python-init", "v3"
+// Output: "gcr.io/datadoghq/dd-lib-python-init@sha256:abc123...", true
+// If resolution fails or is not available, it returns nil.
+// Output: nil, false
+func (r *remoteConfigImageResolver) Resolve(registry string, repository string, tag string) (*ResolvedImage, bool) {
+	if !isDatadoghqRegistry(registry) {
+		log.Debugf("Not a Datadoghq registry, not resolving")
+		return nil, false
+	}
+
+	r.mu.RLock()
+	defer r.mu.RUnlock()
+
+	if len(r.imageMappings) == 0 {
+		log.Debugf("Cache empty, no resolution available")
+		return nil, false
+	}
+
+	repoCache, exists := r.imageMappings[repository]
+	if !exists {
+		log.Debugf("No mapping found for repository %s", repository)
+		return nil, false
+	}
+
+	resolved, exists := repoCache[tag]
+	if !exists {
+		log.Debugf("No mapping found for %s:%s", repository, tag)
+		return nil, false
+	}
+
+	log.Debugf("Resolved %s:%s -> %s", repository, tag, resolved)
+	return &resolved, true
+}
+
+func isDatadoghqRegistry(registry string) bool {
+	var datadoghqRegistries = []string{
+		"gcr.io/datadoghq",
```

Comment:
Here is the full list of registries: https://docs.datadoghq.com/containers/guide/changing_container_registry/